Parsing error when using [] rather than . to parse nested json object

When attempting to evaluate an expression containing a json object using [attributeName] rather than .attributeName notation, I receive the following error

```python
E           lark.exceptions.UnexpectedCharacters: No terminal matches '[' in the current parser context, at line 1 col 81
E
E           pelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]
E                                                   ^
E           Expected one of: 
E               * EXPRESSION_ARRAY_INDEX
E               * EXPRESSION_NULL_CONDITIONAL_OPERATOR
E               * EXPRESSION_WS
E               * DOT
```

In this case, the expression to be evaluated is 
`"@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]"`

which is used to retrieve a specific component of the following json object:
```json
{
    "attribute1": {
        "param1": "value of attribute1, param1",
        "param2": "value of attribute1, param2"
    },
    "attribute2": "value of attribute 2"
}
```

A pipeline parameter is used to specify which component within attribute1 should be returned so this cannot be specified directly.

For the above json object and providing pipeline parameter of parameter = "param1", the expression should evaluate to "value of attribute1, param1".

The expression is evaluated correctly within ADF.

The following expression is also evaluated without error
`"@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1.param1"`

<details><summary>failing unit test</summary>

```python
from data_factory_testing_framework import TestFramework, TestFrameworkType
from data_factory_testing_framework.state import (
    RunParameter,
    RunParameterType,
    DependencyCondition
)
def test():
    testFramework = TestFramework(framework_type=TestFrameworkType.DataFactory, root_folder_path='', should_evaluate_child_pipelines=False,)
    pipeline = testFramework.get_pipeline_by_name(name="test_parse_nested_objects") 
    activities = testFramework.evaluate_pipeline(pipeline, parameters=
            [
                RunParameter(RunParameterType.Pipeline, "parameter", "param1"),
                RunParameter(RunParameterType.System, "pipelineReturnValue", "")
            ]
        )
    
    # activity 1: get the nested object to be consumed by a subsequent activity
    activity = next(activities)
    activity.set_result(DependencyCondition.SUCCEEDED, output = {
        "pipelineReturnValue": {
            "returnValue" :{
                "attribute1": {
                    "param1": "value of attribute1, param1",
                    "param2": "value of attribute1, param2"
                },
                "attribute2": "value of attribute 2"
		    }
        }
    })
    
    # activity 2: read one of the attributes out of the nested response - fails when attempting to evaluate next activity
    activity = next(activities)
```
</details>

<details><summary>Pipelines to reproduce error</summary>
Parent

```json
{
	"name": "test_parse_nested_objects",
	"properties": {
		"activities": [
			{
				"name": "get nested object",
				"type": "ExecutePipeline",
				"dependsOn": [],
				"policy": {
					"secureInput": false
				},
				"userProperties": [],
				"typeProperties": {
					"pipeline": {
						"referenceName": "test_parse_nested_objects_child",
						"type": "PipelineReference"
					},
					"waitOnCompletion": true
				}
			},
			{
				"name": "read value for parameter",
				"type": "SetVariable",
				"dependsOn": [
					{
						"activity": "get nested object",
						"dependencyConditions": [
							"Succeeded"
						]
					}
				],
				"policy": {
					"secureOutput": false,
					"secureInput": false
				},
				"userProperties": [],
				"typeProperties": {
					"variableName": "set_value",
					"value": {
						"value": "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]",
						"type": "Expression"
					}
				}
			}
		],
		"parameters": {
			"parameter": {
				"type": "string",
				"defaultValue": "param1"
			}
		},
		"variables": {
			"nested_config": {
				"type": "String"
			},
			"set_value": {
				"type": "String"
			}
		},
		"annotations": []
	}
}
```

Child

```json
{
	"name": "test_parse_nested_objects_child",
	"properties": {
		"activities": [
			{
				"name": "lookup",
				"type": "SetVariable",
				"dependsOn": [],
				"policy": {
					"secureOutput": false,
					"secureInput": false
				},
				"userProperties": [],
				"typeProperties": {
					"variableName": "pipelineReturnValue",
					"value": [
						{
							"key": "returnValue",
							"value": {
								"type": "Object",
								"content": [
									{
										"key": "attribute1",
										"value": {
											"type": "Object",
											"content": [
												{
													"key": "param1",
													"value": {
														"type": "String",
														"content": "value of attribute1, param1"
													}
												},
												{
													"key": "param2",
													"value": {
														"type": "String",
														"content": "value of attribute1, param2"
													}
												}
											]
										}
									},
									{
										"key": "attribute2",
										"value": {
											"type": "String",
											"content": "value of attribute 2"
										}
									}
								]
							}
						}
					],
					"setSystemVariable": true
				}
			}
		],
		"variables": {
			"nested_config": {
				"type": "String"
			},
			"set_value": {
				"type": "String"
			}
		},
		"annotations": []
	}
}
```
</details>

--- a/src/data_factory_testing_framework/_expression_runtime/data_factory_expression/expression_transformer.py
+++ b/src/data_factory_testing_framework/_expression_runtime/data_factory_expression/expression_transformer.py
@@ -57,7 +57,7 @@ class ExpressionTransformer:
             expression_item_reference: "item" "()"
             expression_pipeline_reference: "pipeline" "()" EXPRESSION_NULL_CONDITIONAL_OPERATOR? "." EXPRESSION_PIPELINE_PROPERTY
             expression_field_reference: "." EXPRESSION_PARAMETER_NAME
-            expression_object_or_array_accessor: "[" expression_evaluation "]" | EXPRESSION_ARRAY_INDEX  // resolves array[0] or object['key'] access
+            expression_object_or_array_accessor: "[" expression_evaluation "]" | EXPRESSION_ARRAY_INDEX | EXPRESSION_ARRAY_OBJECT_ACCESSOR  // resolves array[0], array["key"] or array[<expression>]
 
             // branch rules
             expression_logical_bool: EXPRESSION_LOGICAL_BOOL "(" expression_parameter "," expression_parameter ")"
@@ -72,6 +72,7 @@ class ExpressionTransformer:
             EXPRESSION_DATAFACTORY_REFERENCE: "dataset" | "linkedService"
             EXPRESSION_ACTIVITY_NAME: "'" /[^']*/ "'"
             EXPRESSION_ARRAY_INDEX: ARRAY_INDEX
+            EXPRESSION_ARRAY_OBJECT_ACCESSOR: "[" SINGLE_QUOTED_STRING "]"
             EXPRESSION_BOOLEAN: BOOLEAN
             EXPRESSION_FLOAT: SIGNED_FLOAT
             EXPRESSION_FUNCTION_NAME: {self._supported_functions()}

--- a/src/data_factory_testing_framework/_expression_runtime/data_factory_expression/expression_transformer.py
+++ b/src/data_factory_testing_framework/_expression_runtime/data_factory_expression/expression_transformer.py
@@ -40,7 +40,7 @@ class ExpressionTransformer:
 
         expression_grammar = f"""
             expression_start: "@" expression_evaluation
-            expression_evaluation: (expression_logical_bool | expression_branch | expression_call) ((EXPRESSION_NULL_CONDITIONAL_OPERATOR? (expression_field_reference | nested_expression_array))+)? EXPRESSION_WS*
+            expression_evaluation: (expression_logical_bool | expression_branch | expression_call) ((EXPRESSION_NULL_CONDITIONAL_OPERATOR? (expression_field_reference | expression_object_or_array_accessor))+)? EXPRESSION_WS*
             ?expression_call: expression_function_call
                               // used to translate to expression_pipeline_reference
                               | expression_datafactory_parameters_reference
@@ -49,7 +49,6 @@ class ExpressionTransformer:
                               | expression_variable_reference
                               | expression_item_reference
                               | expression_datafactory_activity_reference
-            nested_expression_array: "[" expression_evaluation "]" | EXPRESSION_ARRAY_INDEX
 
             // reference rules:
             expression_variable_reference: "variables" "(" EXPRESSION_VARIABLE_NAME ")"
@@ -58,6 +57,7 @@ class ExpressionTransformer:
             expression_item_reference: "item" "()"
             expression_pipeline_reference: "pipeline" "()" EXPRESSION_NULL_CONDITIONAL_OPERATOR? "." EXPRESSION_PIPELINE_PROPERTY
             expression_field_reference: "." EXPRESSION_PARAMETER_NAME
+            expression_object_or_array_accessor: "[" expression_evaluation "]" | EXPRESSION_ARRAY_INDEX  // resolves array[0] or object['key'] access
 
             // branch rules
             expression_logical_bool: EXPRESSION_LOGICAL_BOOL "(" expression_parameter "," expression_parameter ")"

--- a/src/data_factory_testing_framework/_expression_runtime/data_factory_expression/expression_transformer.py
+++ b/src/data_factory_testing_framework/_expression_runtime/data_factory_expression/expression_transformer.py
@@ -40,7 +40,7 @@ class ExpressionTransformer:
 
         expression_grammar = f"""
             expression_start: "@" expression_evaluation
-            expression_evaluation: (expression_logical_bool | expression_branch | expression_call) ((EXPRESSION_NULL_CONDITIONAL_OPERATOR? (expression_field_reference | EXPRESSION_ARRAY_INDEX))+)? EXPRESSION_WS*
+            expression_evaluation: (expression_logical_bool | expression_branch | expression_call) ((EXPRESSION_NULL_CONDITIONAL_OPERATOR? (expression_field_reference | nested_expression_array))+)? EXPRESSION_WS*
             ?expression_call: expression_function_call
                               // used to translate to expression_pipeline_reference
                               | expression_datafactory_parameters_reference
@@ -49,6 +49,7 @@ class ExpressionTransformer:
                               | expression_variable_reference
                               | expression_item_reference
                               | expression_datafactory_activity_reference
+            nested_expression_array: "[" expression_evaluation "]" | EXPRESSION_ARRAY_INDEX
 
             // reference rules:
             expression_variable_reference: "variables" "(" EXPRESSION_VARIABLE_NAME ")"

--- a/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
+++ b/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
@@ -327,6 +327,34 @@ from pytest import param as p
             "variable_test_valueparameter_test_value",
             id="variable_and_parameter_reference",
         ),
+        p(
+            "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]",
+            PipelineRunState(
+                parameters=[
+                    RunParameter(RunParameterType.Pipeline, "parameter", "param1"),
+                ],
+                activity_results=[
+                    ActivityResult(
+                        activity_name="get nested object",
+                        status=DependencyCondition.SUCCEEDED,
+                        output={
+                            "pipelineReturnValue": {
+                                "returnValue": {
+                                    "attribute1": [
+                                        {
+                                            "param1": "value1",
+                                        }
+                                    ],
+                                }
+                            }
+                        },
+                    )
+                ]
+            ),
+            "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]",
+            "value1",
+            id="activity_reference_with_nested_property_and_string_array_index",
+        )
     ],
 )
 def test_evaluate(

--- a/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
+++ b/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
@@ -340,11 +340,9 @@ from pytest import param as p
                         output={
                             "pipelineReturnValue": {
                                 "returnValue": {
-                                    "attribute1": [
-                                        {
+                                    "attribute1": {
                                             "param1": "value1",
-                                        }
-                                    ],
+                                        },
                                 }
                             }
                         },

--- a/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
+++ b/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
@@ -341,18 +341,18 @@ from pytest import param as p
                             "pipelineReturnValue": {
                                 "returnValue": {
                                     "attribute1": {
-                                            "param1": "value1",
-                                        },
+                                        "param1": "value1",
+                                    },
                                 }
                             }
                         },
                     )
-                ]
+                ],
             ),
             "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]",
             "value1",
             id="activity_reference_with_nested_property_and_string_array_index",
-        )
+        ),
     ],
 )
 def test_evaluate(

--- a/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
+++ b/tests/unit/functions/test_data_factory_testing_framework_expression_evaluator.py
@@ -351,7 +351,33 @@ from pytest import param as p
             ),
             "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1[pipeline().parameters.parameter]",
             "value1",
-            id="activity_reference_with_nested_property_and_string_array_index",
+            id="activity_reference_with_nested_property_and_parameter_reference",
+        ),
+        p(
+            "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1['param1']",
+            PipelineRunState(
+                parameters=[
+                    RunParameter(RunParameterType.Pipeline, "parameter", "param1"),
+                ],
+                activity_results=[
+                    ActivityResult(
+                        activity_name="get nested object",
+                        status=DependencyCondition.SUCCEEDED,
+                        output={
+                            "pipelineReturnValue": {
+                                "returnValue": {
+                                    "attribute1": {
+                                        "param1": "value1",
+                                    },
+                                }
+                            }
+                        },
+                    )
+                ],
+            ),
+            "@activity('get nested object').output.pipelineReturnValue.returnValue.attribute1['param1']",
+            "value1",
+            id="activity_reference_with_nested_property_and_string_reference",
         ),
     ],
 )

